doc/sqtop.1
.\" Copyright (c) 2010 Marco Steinbach
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\" The following requests are required for all man pages.
.Dd September 18, 2010
.\" .Os [OPERATING_SYSTEM] [version/release]
.Dt SQTOP 1
.Sh NAME
.Nm sqtop
.Nd display active client connections for Squid
.Sh SYNOPSIS
.Nm sqtop
.Op Ar options
.Sh DESCRIPTION
.Nm sqtop
displays information about currently active client connections for a Squid proxy
in a convenient way.
.Pp
Without options,
.Nm sqtop
tries to connect to a Squid proxy listening at 127.0.0.1 using port 3128. 
.Sh OPTIONS
.Bl -tag -width indent
.It Fl -help
Display a brief help text.
.It Fl -host Ar host ( Fl h Ar host )
Squid proxy host. Defaults to 127.0.0.1.
.It Fl -port Ar port ( Fl p Ar port )
Squid proxy port. Defaults to 3128.
.It Fl -pass Ar password ( Fl P Ar password )
Squid proxy cachemgr_passwd.
.It Fl -hosts Ar hostlist ( Fl H Ar hostlist )
Comma-separated list of client IP addresses (CIDR notation is supported) to query the
Squid proxy for. Hostnames are silently ignored.
.It Fl -users Ar userlist ( Fl u Ar userlist )
Comma-separated list of Squid usernames to list active connections for.
.It Fl -brief ( Fl b )
Display brief per-connection information, omits URLs.
.It Fl -detail ( Fl d )
Display detailed information (size, username and average speed) for each URL in each connection.
.It Fl -full ( Fl f )
Display full details (size, username, average speed, delay pool and elapsed time) for each URL in each connection.
.It Fl -zero ( Fl z )
Display zero values instead of silently omitting them.
.It Fl c
Don't compact the display of multiple occurrences of the same URL in a single connection.
.It Fl Z
Don't strip domain part of username.
.It Fl -once ( Fl o )
Disable interactive mode, just print statistics once to stdout.
.It Fl -refreshinterval Ar seconds ( Fl r Ar seconds )
Set the refresh-interval for interactive mode.
.It Fl n
Don't do hostname lookups.
.It Fl S
Don't strip domain part of hostname.
.El
.Sh INTERACTIVE MODE
If built with support for
.Xr ncurses 3 , 
.Nm
defaults to running in interactive mode, occupying the whole screen, unless the
.Ic --once
option was specified on the command line.
.Pp
Information about the Squid server currently connected to, the version of
.Nm
used, as well as eventual error messages are shown at the top of the display.
.Pp
The bottom of the display keeps various aggregates, including current and average speed, the total number of hosts connected and the total number of connections.
.Pp
Any option given on the command line can be changed from within interactive mode by pressing the key corresponding to its respective short option character.
.Pp
In addition to the options given on the command line,
.Nm
recognizes the following keys when in interactive mode:
.Bl -tag -width Fl
.It Ic /
Search for literal substrings in hosts, usernames or URLs.  Regular expressions are not parsed, currently.
.It Ic <space>
Stop refreshing.
.It Ic UP/DOWN, PGUP/PGDN, HOME/END
Scroll display.
.It Ic <enter>
Toggle displayed level of detail for the currently selected entry.
.It Ic \&?
Display the help screen, including current settings for options, where applicable.
.It Ic C
Compact long urls to fit them on one line.
.It Ic s
Toggle mode of display for speed detail between current and average, current only and average only.
.It Ic o
Toggle connection sort order between size, current speed, average speed and max time.
.It Ic R
Toggle hosts showing mode between host name only, host ip only, both ip and host name.
.It Ic q
Quit
.Nm
.El
.Sh EXAMPLES
.Pp
List all currently active connections in interacive mode from 192.168.2.0/24 and 172.18.118.10 to a Squid proxy running at
mysquid.example.com, port 8080 using ZePasswd as cachemgr_passwd:
.Dl $ sqtop -h mysquid.example.com -port 8080 -p ZePasswd -H 192.168.2.0/24,172.18.118.10
.Sh DIAGNOSTICS
.Ex -std sqtop
.Sh SEE ALSO
The Squid homepage at:
.Pa http://www.squid-cache.org .
.Pp
The
.Nm sqtop
homepage at:
.Pa https://github.com/paleg/sqtop .
.Sh AUTHOR
.Nm sqtop
was written by
.An Oleg Palij <o.palij@gmail.com> .
.Pp
This man page was originally written by
.An Marco Steinbach <coco@executive-computing.de> .


debian/changelog
sqtop (2015-02-08) unstable; urgency=low

  * New version 2015-02-08

 -- Oleg Palij <o.palij@gmail.com>  Sun, 08 Feb 2015 17:15:47 +0200

sqtop (2013-12-17) unstable; urgency=low

  * New version 2013-12-17

 -- Oleg Palij <o.palij@gmail.com>  Tue, 17 Dec 2013 16:47:48 +0200

sqtop (2011-11-01) unstable; urgency=low

  * New version 2011-11-01

 -- Oleg Palij <o.palij@gmail.com>  Thu, 03 Nov 2011 14:24:42 +0200

sqtop (2010-10-07-1.1) unstable; urgency=low

  * New version

 -- Oleg Palij <o.palij@gmail.com>  Thu, 26 May 2011 14:40:54 +0300

sqtop (2010-10-07-1) unstable; urgency=low

  * Initial release

 -- Oleg Palij <o.palij@gmail.com>  Fri, 22 Oct 2010 21:37:06 +0300



src/sqtop.cpp
/*
 * (C) 2006 Oleg V. Palij <o.palij@gmail.com>
 * Released under the GNU GPL, see the COPYING file in the source distribution for its full text.
 */

#include "config.h"

//sleep
#include <unistd.h>
#include <getopt.h>
#include <cstdlib>
#include <iostream>
//stringstream
#include <sstream>

#ifdef ENABLE_UI
#include <pthread.h>
#endif

#include "strings.hpp"
#include "Utils.hpp"
#include "ncui.hpp"

using std::string;
using std::cout;
using std::cerr;
using std::endl;
using std::vector;

using namespace sqtop;

#pragma GCC diagnostic push
#pragma GCC diagnostic ignored "-Wwrite-strings"
static struct option longopts[] = {
   { "host",               required_argument,   NULL,    'h' },
   { "port",               required_argument,   NULL,    'p' },
   { "hosts",              required_argument,   NULL,    'H' },
   { "users",              required_argument,   NULL,    'u' },
   { "pass",               required_argument,   NULL,    'P' },
   { "help",               no_argument,         NULL,     0  },
   { "brief",              no_argument,         NULL,    'b' },
   { "full",               no_argument,         NULL,    'f' },
   { "zero",               no_argument,         NULL,    'z' },
   { "detail",             no_argument,         NULL,    'd' },
   { NULL,                 no_argument,         NULL,    'Z' },
#ifdef ENABLE_UI
   { "once",               no_argument,         NULL,    'o' },
   { "refreshinterval",    required_argument,   NULL,    'r' },
#endif
   { NULL,                 no_argument,         NULL,    'c' },
#ifdef WITH_RESOLVER
   { NULL,                 no_argument,         NULL,    'n' },
   { NULL,                 no_argument,         NULL,    'S' },
#endif
   { NULL,                 0,                   NULL,     0 }
};
#pragma GCC diagnostic pop

void usage(char* argv) {
   cout << title << endl;
   cout << "version " << VERSION << " " << copyright << " (" << contacts << ")" << endl;
   cout << endl;
   cout << "Usage:";
   cout << "\n" << argv << " [--help] [--host host] [--port port] [--pass password] [--hosts host1,host...] [--users user1,user2] [--brief] [--detail] [--full] [--zero] [-c] [-Z]";
#ifdef ENABLE_UI
   cout << " [--once] [-r seconds]";
#endif
#ifdef WITH_RESOLVER
   cout << " [-n] [-S]";
#endif
   cout << "\n\t--help                       - show this help;";
   cout << "\n\t--host   (-h) host           - " << host_help << ". Default - '127.0.0.1';";
   cout << "\n\t--port   (-p) port           - " << port_help << ". Default - '3128';";
   cout << "\n\t--pass   (-P) password       - " << passwd_help << ";";
   cout << "\n\t--hosts  (-H) host1,host2... - " << hosts_help << ";";
   cout << "\n\t--users  (-u) user1,user2... - " << users_help << ";";
   cout << "\n\t--brief  (-b)                - " << brief_help << ";";
   cout << "\n\t--detail (-d)                - " << detail_help << ";";
   cout << "\n\t--full   (-f)                - " << full_help << ";";
   cout << "\n\t--zero   (-z)                - " << zero_help << ";";
   cout << "\n\t-c                           - do not " << compact_same_help << ";";
   cout << "\n\t-Z                           - do not " << strip_user_domain_help << ";";
#ifdef ENABLE_UI
   cout << "\n\t--once   (-o)                - disable interactive mode, just print statistics once to stdout;";
   cout << "\n\t--refreshinterval (-r) sec   - " << refresh_interval_help << ";";
#endif
#ifdef WITH_RESOLVER
   cout << "\n\t-n                           - do not " << dns_resolution_help << ";";
   cout << "\n\t-S                           - do not " << strip_host_domain_help << ".";
#endif
   cout << endl;
}

/*static void finish(int sig) {
   foad = sig;
}*/

bool DESC(int a, int b) {
   return a > b;
}

string conns_format(Options* pOpts, vector<SquidConnection> conns) {
   std::stringstream result;

   if (pOpts->compactsameurls)
      sqstat::CompactSameUrls(conns);

   for (vector<SquidConnection>::iterator it = conns.begin(); it != conns.end(); ++it) {
      if (((pOpts->Hosts.size() == 0) || Utils::IPMemberOf(pOpts->Hosts, it->peer)) &&
         ((pOpts->Users.size() == 0) || Utils::UserMemberOf(pOpts->Users, it->usernames))) {

         result << sqstat::ConnFormat(pOpts, *it);

         if (not pOpts->brief) {
            result << endl;
            for (vector<UriStats>::iterator itu = it->stats.begin(); itu != it->stats.end(); ++itu) {
               result << sqstat::StatFormat(pOpts, *it, *itu);
               result << endl;
            }
         }
      result << endl;
      }
   }
   return result.str();
}

#ifdef ENABLE_UI
struct thread_args {
   ncui* ui;
   Options* pOpts;
   sqstat *pSqstat;
};

void squid_loop(void* threadarg) {
   thread_args* pArgs = reinterpret_cast<thread_args*>(threadarg);
   while (true) {
      if (pArgs->pOpts->do_refresh) {
         try {
            pArgs->ui->SetStat( pArgs->pSqstat->GetInfo() );
            pArgs->ui->ClearError();
         }
         catch (sqstatException &e) {
            pArgs->ui->SetError(e.what());
         }
         pArgs->ui->Tick();
      }
      for (int i=0; i<pArgs->pOpts->sleep_sec; ++i) {
         sleep(1);
      }
   }
}
#endif

int main(int argc, char **argv) {
   // TODO: config file ?
   int ch;
   string tempusers;

   sqtop::Options* pOpts = new Options();

   string getopt_options = "u:H:h:p:P:dzbfc";
#ifdef ENABLE_UI
   getopt_options += "r:o";
#endif
#ifdef WITH_RESOLVER
   getopt_options += "nS";
#endif
   while ((ch = getopt_long(argc, argv, getopt_options.c_str(), longopts, NULL)) != -1) {
      switch (ch) {
         case 'd':
            pOpts->detail = true;
            break;
         case 'z':
            pOpts->zero = true;
            break;
         case 'b':
            pOpts->brief = true;
            break;
         case 'f':
            pOpts->detail = true;
            pOpts->full = true;
            break;
         case 'P':
            pOpts->pass = optarg;
            break;
         case 'h':
            pOpts->host = optarg;
            break;
         case 'p':
            try {
               pOpts->port = Utils::stol(optarg);
            }
            catch (const std::exception& error) {
               cerr << "Wrong port - '" << optarg << "' (" << error.what() << ")" << endl;
               exit(1);
            }
            break;
         case 'H':
            pOpts->Hosts = Utils::SplitString(optarg, ",");
            break;
         case 'u':
            tempusers = optarg;
            Utils::ToLower(tempusers);
            pOpts->Users = Utils::SplitString(tempusers, ",");
            break;
#ifdef ENABLE_UI
         case 'o':
            pOpts->ui = false;
            break;
         case 'r':
            try {
               long int sec = Utils::stol(optarg);
               if (sec > 0) {
                  pOpts->sleep_sec = sec;
               } else {
                  cerr << "Refresh interval should be greater than 0, using default - " << pOpts->sleep_sec << endl;
               }
            }
            catch (const std::exception& error) {
               cerr << "Wrong refresh interval - '" << optarg << "' (" << error.what() << ")" << endl;
               exit(1);
            }
            break;
#endif
#ifdef WITH_RESOLVER
         case 'n':
            pOpts->dns_resolution = false;
            break;
         case 'S':
            pOpts->strip_host_domain = false;
            break;
#endif
         case 'c':
            pOpts->compactsameurls = false;
            break;
         case 'Z':
            pOpts->strip_user_domain = false;
            break;
         default:
            usage(argv[0]);
            exit(0);
      }
   }

   // TODO:
   //struct sigaction sa = {};
   //sa.sa_handler = finish;
   //sigaction(SIGINT, &sa, NULL);

   sqstat *pSqstat;

#ifdef WITH_RESOLVER
   Resolver *pResolver = new Resolver();
   pSqstat = new sqstat(pOpts, pResolver);
#else
   pSqstat = new sqstat(pOpts);
#endif


#ifdef ENABLE_UI
   if (pOpts->ui) {
#ifdef WITH_RESOLVER
      ncui *ui = new ncui(pOpts, pResolver);
#else
      ncui *ui = new ncui(pOpts);
#endif
      ui->CursesInit();
      thread_args args;
      args.ui = ui;
      args.pOpts = pOpts;
      args.pSqstat = pSqstat;
#ifdef WITH_RESOLVER
      pResolver->Start(MAX_THREADS);
      pResolver->resolve_mode = Resolver::RESOLVE_ASYNC;
#endif

      pthread_t sq_thread;
      pthread_create(&sq_thread, NULL, (void *(*) (void *)) &squid_loop, (void *) &args);

      ui->Loop();

      pthread_cancel(sq_thread);
      ui->CursesFinish();
      delete ui;
   } else {
#endif
      SquidStats sqstats;
#ifdef WITH_RESOLVER
      pResolver->resolve_mode = Resolver::RESOLVE_SYNC;
#endif
      pOpts->speed_mode = Options::SPEED_AVERAGE;
      try {
         sqstats = pSqstat->GetInfo();
      }
      catch (sqstatException &e) {
         cerr << e.what() << endl;
         exit(1);
      }
      cout << sqstat::HeadFormat(pOpts, sqstats.total_connections, sqstats.connections.size(), sqstats.av_speed) << endl;
      cout << conns_format(pOpts, sqstats.connections) << endl;
#ifdef ENABLE_UI
   }
#endif

   delete pSqstat;
#ifdef WITH_RESOLVER
   delete pResolver;
#endif
   delete pOpts;

   return 0;
}

// vim: ai ts=3 sts=3 et sw=3 expandtab
